### DISCUSSION TITLE
submodule compilation

### DISCUSSION THREAD

**Discussion by danielhrisca** [danielhrisca/asammdf#1121] [Q&A] [answered]
Hello @zariiii9003 ,

I need your support to compile a submodule. 

In the branch https://github.com/danielhrisca/asammdf/tree/zlib I'm trying to use the isa-l library in the C extension to decompress the data blocks (the code looks awful but I'm just in the POC phase) . I don't know how to configure the build system to include the header files correctly for the submodule.

Do you have any idea how to do it?

**Comment by zariiii9003** [comment]
I'll try :)

**Reply by zariiii9003** [reply]
@danielhrisca I made some progress on WIndows, but there seem to be a few [syntax errors](https://github.com/zariiii9003/asammdf/actions/runs/12371788317/job/34528633987#step:7:418) in zlib_isal.c.

You need [NASM](https://www.nasm.us/) in your path to compile ISAL.
Here is my [WIP branch](https://github.com/zariiii9003/asammdf/tree/my_zlib)

**Reply by zariiii9003** [reply]
@danielhrisca I think this should work once the code errors are fixed: https://github.com/zariiii9003/asammdf/commit/b51ce9048a6c2ec37546d1e906dce66db3b62824

**Reply by danielhrisca** [reply]
thank you @zariiii9003 ! I appreciate your effort!

**Reply by zariiii9003** [reply]
No problem. Ping me, if you need help with CMake and libdeflate.

**Reply by danielhrisca** [reply]
I could not get isal to work, and on a further check libdeflate actually performs faster for decompression (at least with my data). 

Now there are two issues:
* using `__cdecl` on linux throws a compilation error
* I've done a dynamic linking to the libdeflate dll which I tested to be working ok on Windows. I'm not sure if the code runs ok on linux. Do you think it makes better sense to libdefalte statically linked as a submodule?

**Reply by zariiii9003** [reply]
I think static linking will be less complicated. Otherwise you'd need to somehow include the correct file (.dll, .so, .dylib) in the wheel and find it at runtime.

**Reply by zariiii9003** [reply]
Since libdeflate uses CMake, it should be quite easy to setup: https://github.com/zariiii9003/asammdf/commit/38a7d1e29d95249bef83196f32a598de19cffa26

I added libdeflate as a submodule in ext/libdeflate

**Reply by danielhrisca** [reply]
The compilation fails on my PC

```
python -m build --no-isolation
* Getting build dependencies for sdist...
* Building sdist...
*** scikit-build-core 0.10.7 (sdist)
* Building wheel from sdist
* Getting build dependencies for wheel...
* Building wheel...
*** scikit-build-core 0.10.7 using CMake 3.31.1 (wheel)
*** Configuring CMake...
2024-12-20 10:36:02,441 - scikit_build_core - WARNING - Can't find a Python library, got libdir=None, ldlibrary=None, multiarch=None, masd=None
2024-12-20 10:36:02,441 - scikit_build_core - WARNING - Can't find a Python library, got libdir=None, ldlibrary=None, multiarch=None, masd=None
loading initial cache file C:\Users\usr\AppData\Local\Temp\tmpo4747dsm\build\CMakeInit.txt
-- Building for: Visual Studio 17 2022
-- Selecting Windows SDK version 10.0.22621.0 to target Windows 10.0.22631.
-- The C compiler identification is MSVC 19.41.34120.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2022/BuildTools/VC/Tools/MSVC/14.41.34120/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Found Python: C:\Users\usr\01__Workspace\__venv_cp312\Scripts\python.exe (found version "3.12.4") found components: Interpreter Development.Module Development.SABIModule NumPy
CMake Deprecation Warning at ext/libdeflate/CMakeLists.txt:1 (cmake_minimum_required):
  Compatibility with CMake < 3.10 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value.  Or, use the <min>...<max> syntax
  to tell CMake that the project requires at least <min> but has been updated
  to work with policies introduced by <max> or earlier.


-- Looking for clock_gettime
-- Looking for clock_gettime - not found
-- Looking for futimens
-- Looking for futimens - not found
-- Looking for posix_fadvise
-- Looking for posix_fadvise - not found
-- Looking for posix_madvise
-- Looking for posix_madvise - not found
-- Performing Test HAVE_STAT_NANOSECOND_PRECISION
-- Performing Test HAVE_STAT_NANOSECOND_PRECISION - Failed
-- Configuring done (7.9s)
CMake Error at ext/libdeflate/CMakeLists.txt:247 (add_library):
  Cannot find source file:

    lib/arm/cpu_features.c

  Tried extensions .c .C .c++ .cc .cpp .cxx .cu .mpp .m .M .mm .ixx .cppm
  .ccm .cxxm .c++m .h .hh .h++ .hm .hpp .hxx .in .txx .f .F .for .f77 .f90
  .f95 .f03 .hip .ispc


CMake Error at ext/libdeflate/CMakeLists.txt:247 (add_library):
  No SOURCES given to target: libdeflate_shared


CMake Generate step failed.  Build files cannot be regenerated correctly.

*** CMake configuration failed

ERROR Backend subprocess exited when trying to invoke build_wheel
```

**Reply by zariiii9003** [reply]
@danielhrisca Your branch compiles in-source with this command `pip install --no-build-isolation --config-settings=editable.rebuild=true -Cbuild-dir=build -ve.`

But normal builds fail with the same error for me. I noticed, that `ext/libdeflate/lib` is missing in the sdist. Maybe there's some gitignore rule that causes this...

**Reply by zariiii9003** [reply]
Changing `lib/` to `/lib/` in `.gitignore` solves this.

**Reply by danielhrisca** [reply]
Awesome, thanks a lot!

**Reply by zariiii9003** [reply]
Another thing i noticed: There's a bunch of libdeflate build artifacts in the wheel. Adding `EXCLUDE_FROM_ALL` in `add_subdirectory(ext/libdeflate EXCLUDE_FROM_ALL)` seems to fix that.

**Reply by danielhrisca** [reply]
So far so good, I think there is a last step to have it building on linux 
https://github.com/danielhrisca/asammdf/actions/runs/12431138047/job/34707963781

**Reply by zariiii9003** [reply]
Try adding `set(CMAKE_POSITION_INDEPENDENT_CODE ON)`

**Reply by danielhrisca** [reply]
That worked! Once again thanks, I have 0 knowledge of CMAKE or complicated C projects compilation in general.

**Reply by danielhrisca** [reply]
To get an idea of what this hopefully will lead to: 

I have demo MF4 file of 32GB which contains uncompressed data of ~360GB. All the 44000 signals are found in a single channel group within the file, which means accessing a signal requires to process all the uncompressed data.

With the new C code I was able to extract 200 signals from the total of 44000 in about 70s when previously this took more than 8 minutes

**Reply by zariiii9003** [reply]
> That worked! Once again thanks, I have 0 knowledge of CMAKE or complicated C projects compilation in general.

Great! I also have very basic knowledge of CMake tbh 😄
But it's certainly easier to have direct access to the build system than going through setup.py.

You should add this to be compliant with the libdeflate license:
```
# Copy 3rd party license files
install(FILES "${CMAKE_SOURCE_DIR}/ext/libdeflate/COPYING"
        DESTINATION "${SKBUILD_METADATA_DIR}/licenses"
        RENAME LIBDEFLATE_LICENSE)
```

**Reply by danielhrisca** [reply]
Hello @zariiii9003 @a-detiste 

I'm trying to add [lz4](https://github.com/lz4/lz4) as submodule, but I'm failing miserably. Can you help with this? Thanks!

see (https://github.com/danielhrisca/asammdf/tree/lz4_c)

**Reply by zariiii9003** [reply]
@danielhrisca i think the submodule was not properly added, because the path is not found in the CI pipeline:
 `add_subdirectory given source "ext/lz4/build/cmake" which is not an existing directory.` 

Try `git submodule add https://github.com/lz4/lz4 ext/lz4`, then we'll check the next error.

**Reply by zariiii9003** [reply, marked as answer]
@danielhrisca https://github.com/zariiii9003/asammdf/commit/d0edce241ed2bbc7001bd9c1c84661bc3e4aab45

**Reply by danielhrisca** [reply]
thanks that works !
